pkg/kinflate/transformers/overlay.go

```diff
@@ -59,39 +65,143 @@ func (o *overlayTransformer) Transform(baseResourceMap resource.ResourceCollecti
 		switch {
 		case runtime.IsNotRegisteredError(err):
 			// Use JSON merge patch to handle types w/o schema
-			baseBytes, err := json.Marshal(base.Data)
+			merged, err = jsonMergePatch(base.Data.Object, overlay.Data.Object, jsonpatch.MergePatch)
 			if err != nil {
 				return err
 			}
-			patchBytes, err := json.Marshal(overlay.Data)
+		case err != nil:
+			return err
+		default:
+			// Use Strategic Merge Patch to handle types w/ schema
+			// TODO: Change this to use the new Merge package.
+			// Store the name of the base object, because this name may have been munged.
+			// Apply this name to the StrategicMergePatched object.
+			lookupPatchMeta, err := strategicpatch.NewPatchMetaFromStruct(versionedObj)
 			if err != nil {
 				return err
 			}
-			mergedBytes, err := jsonpatch.MergePatch(baseBytes, patchBytes)
+			merged, err = strategicpatch.StrategicMergeMapPatchUsingLookupPatchMeta(
+				base.Data.Object,
+				overlay.Data.Object,
+				lookupPatchMeta)
 			if err != nil {
 				return err
 			}
-			err = json.Unmarshal(mergedBytes, &merged)
+		}
+		base.Data.SetName(baseName)
+		baseResourceMap[gvkn].Data.Object = merged
+	}
+	return nil
+}
+
+// mergePatches merge and index patches by GVKN.
+// It errors out if there is conflict between patches.
+func (o *overlayTransformer) mergePatches() (resource.ResourceCollection, error) {
```

Comment:
This function is too long. It's cyclomatic complexity terrifies me.

Reply:
Will fix.